src/realm/sync/noinst/client_impl_base.hpp

```diff
@@ -56,6 +56,57 @@ class SessionWrapperStack {
     SessionWrapper* m_back = nullptr;
 };
 
+/// Register a function whose invocation can be triggered repeatedly.
+///
+/// While the function is always executed by the event loop thread, the
+/// triggering of its execution can be done by any thread.
+///
+/// The function is guaranteed to not be called after the Trigger object is
+/// destroyed.
+///
+/// Note that even though the trigger() function is thread-safe, the Trigger
+/// object, as a whole, is not. In particular, construction and destruction must
+/// not be considered thread-safe.
+class Trigger : public std::enable_shared_from_this<Trigger> {
+public:
+    Trigger(network::Service* service, std::function<void()>&& handler);
+    ~Trigger() noexcept = default;
+
+    /// Trigger another invocation of the associated function.
+    ///
+    /// An invocation of trigger() puts the Trigger object into the triggered
+    /// state. It remains in the triggered state until shortly before the
+    /// function starts to execute. While the Trigger object is in the triggered
+    /// state, trigger() has no effect. This means that the number of executions
+    /// of the function will generally be less that the number of times
+    /// trigger() is invoked.
+    ///
+    /// A particular invocation of trigger() ensures that there will be at least
+    /// one invocation of the associated function whose execution begins after
+    /// the beginning of the execution of trigger(), so long as the event loop
+    /// thread does not exit prematurely from run().
+    ///
+    /// If trigger() is invoked from the event loop thread, the next execution
+    /// of the associated function will not begin until after trigger() returns,
+    /// effectively preventing reentrancy for the associated function.
+    ///
+    /// If trigger() is invoked from another thread, the associated function may
+    /// start to execute before trigger() returns.
+    ///
+    /// Note that the associated function can retrigger itself, i.e., if the
+    /// associated function calls trigger(), then that will lead to another
+    /// invocation of the associated function, but not until the first
+    /// invocation ends (no reentrance).
+    void trigger();
+
+protected:
+    // TODO: replace network::Service with SyncSocketProvider when DefaultSocketProvider is created
+    network::Service* m_service;
```

Comment:
Why don't we make this class templated on the Service type and then we don't need a TODO here.